pkg/autoscaler/autoscaler/util.go

```diff
@@ -126,13 +136,186 @@ func defaultTAC(tac *v1alpha1.TidbClusterAutoScaler) {
+
+	var scaleOutCount int
+	var scaleInCount int
+	scaleOutCount, scaleInCount = 0, 0
+	var err error
+
+	if v, existed := tac.Annotations[targetScaleOutAnn]; existed {
+		scaleOutCount, err = strconv.Atoi(v)
+		if err != nil {
+			return fmt.Errorf(invalidTacAnnotationErrorMsg, tac.Namespace, tac.Name, err)
+		}
+	}
+
+	if v, existed := tac.Annotations[targetScaleInAnn]; existed {
+		scaleInCount, err = strconv.Atoi(v)
+		if err != nil {
+			return fmt.Errorf(invalidTacAnnotationErrorMsg, tac.Namespace, tac.Name, err)
+		}
+	}
+
+	if currentReplicas < recommendedReplicas {
+		// scale-out
+		scaleOutCount = scaleOutCount + 1
+		scaleInCount = 0
+	} else if currentReplicas > recommendedReplicas {
+		// scale-in
+		scaleOutCount = 0
+		scaleInCount = scaleInCount + 1
+	} else {
+		scaleOutCount = 0
+		scaleInCount = 0
+	}
+
+	// update tc annotation
+	tac.Annotations[targetScaleOutAnn] = fmt.Sprintf("%d", scaleOutCount)
+	tac.Annotations[targetScaleInAnn] = fmt.Sprintf("%d", scaleInCount)
+	return nil
+}
+
+func checkConsecutiveCount(tac *v1alpha1.TidbClusterAutoScaler,
+	memberType v1alpha1.MemberType, currentReplicas int32, recommendedReplicas int32) (bool, error) {
+	if currentReplicas == recommendedReplicas {
+		return false, nil
+	}
+	targetScaleOutAnn := fmt.Sprintf("%s.%s", memberType.String(), annScaleOutSuffix)
+	targetScaleInAnn := fmt.Sprintf("%s.%s", memberType.String(), annScaleInSuffix)
+	currentScaleOutCount, err := strconv.ParseInt(tac.Annotations[targetScaleOutAnn], 10, 32)
+	if err != nil {
+		return false, err
+	}
+	currentScaleInCount, err := strconv.ParseInt(tac.Annotations[targetScaleInAnn], 10, 32)
+	if err != nil {
+		return false, err
+	}
+	switch memberType {
+	case v1alpha1.TiDBMemberType:
+		if currentReplicas < recommendedReplicas {
+			// scale-out
+			if int32(currentScaleOutCount) < *tac.Spec.TiDB.ScaleOutThreshold {
+				return false, nil
+			}
+		} else {
+			// scale-in, no-scaling would be return nil at first
+			if int32(currentScaleInCount) < *tac.Spec.TiDB.ScaleInThreshold {
+				return false, nil
+			}
+		}
+	case v1alpha1.TiKVMemberType:
+		if currentReplicas < recommendedReplicas {
+			// scale-out
+			if int32(currentScaleOutCount) < *tac.Spec.TiKV.ScaleOutThreshold {
+				return false, nil
+			}
+		} else {
+			// scale-in, no-scaling would be return nil at first
+			if int32(currentScaleInCount) < *tac.Spec.TiDB.ScaleInThreshold {
+				return false, nil
+			}
+		}
+	default:
+		return false, fmt.Errorf(invalidMemberTypeErrorMsg, tac.Namespace, tac.Name, memberType)
+	}
+	return true, nil
+}
+
+func emptyAutoScalingCountAnn(tac *v1alpha1.TidbClusterAutoScaler, memberType v1alpha1.MemberType) {
+	targetScaleOutAnn := fmt.Sprintf("%s.%s", memberType.String(), annScaleOutSuffix)
+	targetScaleInAnn := fmt.Sprintf("%s.%s", memberType.String(), annScaleInSuffix)
+	if tac.Annotations == nil {
+		tac.Annotations = map[string]string{}
```

Comment:
`return` if nil?

Reply:
I think not necessary,  update annotation value as "0" is also ok.